theano/gpuarray/roi_pool.c

```diff
@@ -0,0 +1,187 @@
+        for (int roi_n = 0; roi_n < num_rois; ++roi_n) {
+            const float* offset_bottom_rois = bottom_rois + roi_n * 5;
+            int roi_batch_index = offset_bottom_rois[0];
+            if (n != roi_batch_index) {
+            continue;
+            }
+
+            int roi_start_w = round(offset_bottom_rois[1] * spatial_scale);
+            int roi_start_h = round(offset_bottom_rois[2] * spatial_scale);
+            int roi_end_w = round(offset_bottom_rois[3] * spatial_scale);
+            int roi_end_h = round(offset_bottom_rois[4] * spatial_scale);
+
+            // Skip if ROI doesn't include (h, w)
+            const bool in_roi = (w >= roi_start_w && w <= roi_end_w &&
+                               h >= roi_start_h && h <= roi_end_h);
+            if (!in_roi) {
+            continue;
+            }
+
+            int offset = (roi_n * channels + c) * pooled_height * pooled_width;
+            const float* offset_top_diff = top_diff + offset;
+            const float* offset_argmax_data = argmax_data + offset;
+
+            // Compute feasible set of pooled units that could have pooled
+            // this bottom unit
+
+            // Force malformed ROIs to be 1x1
+            int roi_width = max(roi_end_w - roi_start_w + 1, 1);
+            int roi_height = max(roi_end_h - roi_start_h + 1, 1);
+
+            float bin_size_h = (float)roi_height
+                             / (float)pooled_height;
+            float bin_size_w = (float)roi_width
+                             / (float)pooled_width;
+
+            int phstart = floor((float)(h - roi_start_h) / bin_size_h);
+            int phend = ceil((float)(h - roi_start_h + 1) / bin_size_h);
+            int pwstart = floor((float)(w - roi_start_w) / bin_size_w);
+            int pwend = ceil((float)(w - roi_start_w + 1) / bin_size_w);
+
+            phstart = min(max(phstart, 0), pooled_height);
+            phend = min(max(phend, 0), pooled_height);
+            pwstart = min(max(pwstart, 0), pooled_width);
+            pwend = min(max(pwend, 0), pooled_width);
+
+            for (int ph = phstart; ph < phend; ++ph) {
+                for (int pw = pwstart; pw < pwend; ++pw) {
+                  if ((int)(offset_argmax_data[ph * pooled_width + pw]) == 
+                      (h * width + w)) {
+                    gradient += offset_top_diff[ph * pooled_width + pw];
+                  }
+                }
+            }
+        }
+        bottom_diff[index] = gradient;
+    }
+}
+
+
+int APPLY_SPECIFIC(GPUBackward)(PyGpuArrayObject* data,
+                           PyGpuArrayObject* rois,
+                           PyGpuArrayObject* out_grad,
+                           PyGpuArrayObject** data_grad) {
+    int count = gpuarray_get_elsize(data);
+    int batch_size = PyGpuArray_DIMS(rois)[0];
+    int channels = PyGpuArray_DIMS(data)[1];
+    int height = PyGpuArray_DIMS(data)[2];
+    int width = PyGpuArray_DIMS(data)[3];
```

Comment:
Similarly to GpuForward, you should perform some checks:
- prepare and check data_grad
- check that input arrays are c-contiguous.